scripts/emulator-testing/emulators/emulator.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { spawn } from 'child-process-promise';
+import { ChildProcess } from 'child_process';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as request from 'request';
+import * as tmp from 'tmp';
+
+export abstract class Emulator {
+  binaryName: string;
+  binaryUrl: string;
+  binaryPath: string;
+
+  emulator: ChildProcess;
+  port: number;
+
+  constructor(port: number) {
+    this.port = port;
+  }
+
+  async download(): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      tmp.dir((err, dir) => {
+        if (err) reject(err);
+
+        console.log(`Created temporary directory at [${dir}].`);
+        const filepath: string = path.resolve(dir, this.binaryName);
+        const writeStream: fs.WriteStream = fs.createWriteStream(filepath);
+
+        console.log(`Downloading emulator from [${this.binaryUrl}] ...`);
+        request(this.binaryUrl)
+          .pipe(writeStream)
+          .on('finish', () => {
+            console.log(`Saved emulator binary file to [${filepath}].`);
+            this.binaryPath = filepath;
+            resolve();
+          })
+          .on('error', reject);
+      });
+    });
+  }
+
+  async setUp(): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      const promise: any = spawn(
```

Comment:
Generally speaking, any is not allowed. Since typing is not available for `child-process-promise`, you can define the typing yourself as 
```javascript 
interface ChildProcessPromise extends Promise<void> {
   childProcess: number;
}
```

Reply:
Done.